### ISSUE TITLE
Vertaisarvionti

### ISSUE THREAD

**Issue by TuuPu** [aaartyomsss/DigitsClassifier_TiRa#1]
11.40 18.5

Note: Wrote this in English, since the project is done in English.

As you have stated in UI and Implementation part, the project still has CNN implemented using scikit-learn. This makes it fairly hard for me to write a peer-review at this point, since we were mostly assigned to peer-review the actual implementation and testing, not so much the UI or the documentation.

That being said, I can appreciate the fact that you have used Docker and the UI is fairly versatile for a project that is mostly based on algorithms and data-structures. The implementation and the possibility for a user to have their own number tested is a really fun and nice feature! Though it makes me wonder if you should have focused more on the CNN at first and leave the UI and other features for later. Since the CNN is the focus of this project.

The tests are done well in parts where they are done. You have made good comments on why testing the speed of the training is on the other hand necessary and on the other hand not something that should be considered that much. The tests considering accuracy and saving are fairly straight forward and clean at this point since the scikit-learn does the work, I assume these tests will get more complicated and testing is going to get more versatile once you start building the project from scratch.
To make the tests more versatile at this point, you could have also written tests to see if a user drawn image is saved in the correct form and also written tests to see that the MNIST data loaded from the csv-files are saved in the correct format. It is quite hard to evaluate the testing overall since coverage is missing from the project.

As for the neural network, it does everything it is supposed to do, but scikit's tools handle the work completely. You've split up the functions in an understandable and clear way and I assume that almost all of the functions will look different in the future when you start building your CNN from a scratch. Once again, I'm struggling to write a peer-review about this since there is not that much to peer-review. But from the looks of the project, you clearly have a good vision of what you want to do and it definitely seems like you know what you are doing since the UI and other features are so complex. What I would have done differently (possibly) is that I would have split up manipulating the data (e.g. loading up the numbers and splitting up the test and training set and such) and the actual CNN-side of the module to two separate modules. But this is more of a taste thing than anything else. Plus on my own project, the MNIST-dataset is more heavily manipulated, so having a module to do that work makes more sense.

Even though I am not supposed to include notes or remarks about UI or documentation, I would have appreciated more documentation within the modules. General docstring descriptions of what the functions are supposed to do would have been great. I understand the lack of documentation in a sense that the modules are not close to being ready, but I assume the structure of the modules and functions are probably going to stay the same, but the "insides" of the functions are likely to change.

I was not completely sure how you are going to build the CNN, since the you have mentioned a whole book as your resource for the implementation, but I would be interested in knowing how it is done. CNN implementation is fairly new to me and I can honestly say that I only know the basics and theory aspects of how a CNN works. My own project also does image recognition but uses K-nearest neighbors for it. So I'd be happy to hear about the implementation from a scratch and I think I'll look into your project later on to see how it evolves!
